pkg/dumper/generic/sql.go

```diff
@@ -0,0 +1,150 @@
+package generic
+
+import (
+	"sync"
+
+	"github.com/hellofresh/klepto/pkg/config"
+	"github.com/hellofresh/klepto/pkg/database"
+	"github.com/hellofresh/klepto/pkg/dumper"
+	"github.com/hellofresh/klepto/pkg/reader"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	sqlDumper struct {
+		SqlEngine
+
+		reader reader.Reader
+	}
+
+	SqlEngine interface {
+		DumpStructure(sql string) error
+
+		DumpTable(tableName string, rowChan <-chan database.Row) error
+
+		// Close closes the dumper resources and releases them.
+		Close() error
+	}
+
+	SqlEngineAdvanced interface {
+		PreDumpTables([]string) error
+		PostDumpTables([]string) error
+	}
+)
+
+func NewSqlDumper(rdr reader.Reader, engine SqlEngine) dumper.Dumper {
+	return &sqlDumper{
+		SqlEngine: engine,
+		reader:    rdr,
+	}
+}
+
+func (p *sqlDumper) Dump(done chan<- struct{}, configTables config.Tables) error {
+	if err := p.readAndDumpStructure(); err != nil {
+		return err
+	}
+
+	return p.readAndDumpTables(done, configTables)
+}
+
+func (p *sqlDumper) readAndDumpStructure() error {
+	log.Debug("Dumping structure...")
+	structureSQL, err := p.reader.GetStructure()
+	if err != nil {
+		return errors.Wrap(err, "failed to get structure")
+	}
+
+	if err := p.DumpStructure(structureSQL); err != nil {
+		return errors.Wrap(err, "failed to dump structure")
+	}
+
+	log.Debug("Structure dumped")
+	return nil
+}
+
+func (p *sqlDumper) readAndDumpTables(done chan<- struct{}, configTables config.Tables) error {
+	// Get the tables
+	tables, err := p.reader.GetTables()
+	if err != nil {
+		return err
+	}
+
+	// Trigger pre dump tables
+	if adv, ok := p.SqlEngine.(SqlEngineAdvanced); ok {
+		if err := adv.PreDumpTables(tables); err != nil {
+			return err
+		}
+	}
+
+	// TODO make the amount on concurrent dumps configurable
+	semaphoreChan := make(chan struct{}, 10)
+
+	var wg sync.WaitGroup
+	wg.Add(len(tables))
+	for _, tbl := range tables {
+		semaphoreChan <- struct{}{}
+
+		var opts reader.ReadTableOpt
+
+		table, err := configTables.FindByName(tbl)
+		if err != nil {
+			log.WithError(err).WithField("table", tbl).Debug("no configuration found for table")
+		}
+
+		if table != nil {
+			opts = reader.ReadTableOpt{
+				Limit:         table.Filter.Limit,
+				Relationships: p.relationshipConfigToOptions(table.Relationships),
+			}
+		}
+
+		// Create read/write chanel
+		rowChan := make(chan database.Row)
+
+		go func(tableName string, rowChan <-chan database.Row) {
+			if err := p.DumpTable(tableName, rowChan); err != nil {
+				log.WithError(err).WithField("table", tableName).Error("Failed to dump table")
+			}
+
+			wg.Done()
+			<-semaphoreChan
+		}(tbl, rowChan)
+
+		go func(tableName string, opts reader.ReadTableOpt, rowChan chan<- database.Row) {
+			if err := p.reader.ReadTable(tableName, rowChan, opts); err != nil {
+				log.WithError(err).WithField("table", tableName).Error("Failed to read table")
+			}
+		}(tbl, opts, rowChan)
+	}
+
+	go func() {
```

Comment:
Given this goroutine is for closing resources, I'd normally declare it above the for range loop:

```go
go func() {
	wg.Wait()
	close(semaphoreChan)
	if adv, ok := p.SqlEngine.(SqlEngineAdvanced); ok {
		if err := adv.PostDumpTables(tables); err != nil {
			log.WithError(err).Error("Post dump tables failed")
		}
	}
	done <- struct{}{}
}()

for _, tbl := range tables {
}
```

However feel free if you think it's better to have it declared where it is, but note the `close(semaphoreChan)` must be called